strawberryfields/gbs/similarity.py

```diff
@@ -201,3 +226,151 @@ def event_to_sample(photon_number: int, max_count_per_mode: int, modes: int) ->
             available_modes.remove(j)
 
     return sample
+
+
+def _product_factorial(l: list) -> int:
+    """Computes the product of the factorial of elements in ``l``.
+
+    Args:
+        l (list[int]): the input list
+
+    Returns:
+        int: the product of factorials
+    """
+    return np.prod(factorial(l))
+
+
+def orbit_cardinality(orbit: list, modes: int) -> int:
+    """Gives the number of samples belonging to the input orbit.
+
+    **Example usage**:
+
+    >>> orbit_cardinality([2, 1, 1], 4)
+    12
+
+    Args:
+        orbit (list[int]): orbit to count number of samples
+        modes (int): number of modes in the sample
+
+    Returns:
+        int: number of samples in the orbit
+    """
+    sample = _padded_orbit(orbit, modes)
+    counts = list(Counter(sample).values())
+    return int(factorial(modes) / _product_factorial(counts))
+
+
+def event_cardinality(photon_number: int, max_count_per_mode: int, modes: int) -> int:
+    """Gives the number of samples belonging to the input event.
+
+    **Example usage**:
+
+    >>> event_cardinality(5, 2, 6)
+    126
+
+    Args:
+        photon_number (int): number of photons in the event
+        max_count_per_mode (int): maximum number of photons per mode in the event
+        modes (int): number of modes in the samples
+
+    Returns:
+        int: number of samples in the event
+    """
+    cardinality = 0
+
+    for orb in orbits(photon_number):
+        if max(orb) <= max_count_per_mode:
+            cardinality += orbit_cardinality(orb, modes)
+
+    return cardinality
+
+
+def p_orbit_mc(orbit: list, graph: nx.Graph, n_mean: float = 5, samples: int = 1000) -> float:
+    """Gives a Monte Carlo estimate of the probability of a given orbit with respect to a graph.
+
+    To make this estimate, several samples from the orbit are drawn uniformly at random using
+    :func:`orbit_to_sample`.
+
+    For each sample, this function calculates the probability of observing that sample from a GBS
+    device programmed according to the input graph and mean photon number. Calculating sample
+    probabilities requires computing normalization constants and the hafnian of the subgraph
+    specified by the sample :cite:`hamilton2017`.
+
+    The sum of the probabilities is then rescaled according to the cardinality of the
+    orbit and the total number of samples. The resultant estimate has sample mean equal to the
+    orbit probability.
+
+    **Example usage**:
+
+    >>> graph = nx.complete_graph(10)
+    >>> p_orbit_mc([2, 1, 1], graph)
+    0.5
+
+    Args:
+        orbit (list[int]): orbit for which to estimate the probability
+        graph (nx.Graph): input graph encoded in the GBS device
+        n_mean (float): total mean photon number of the GBS device
+        samples (int): number of samples used in the Monte Carlo estimation
+
+    Returns:
+        float: estimated probability of the orbit with respect to the graph
+    """
+    modes = graph.order()

```

Comment:
I feel like `p_orbit_mc` and `estimate_event_prob` could harness SF rather than calling the more elementary parts of `thewalrus`. E.g., we could call `state.fock_prob([1,2,0,1])`? (@josh146)